document-store/src/main/java/org/hypertrace/core/documentstore/query/Query.java

```diff
@@ -61,227 +48,8 @@
  *         .build();
  *  </code>
  */
-@AllArgsConstructor(access = AccessLevel.PRIVATE)
-public final class Query {
-  private final Selection selection; // Missing selection represents fetching all the columns
-  private final Filter filter;
-
-  private final Aggregation aggregation;
-  private final Filter aggregationFilter;
-
-  private final Sort sort;
-  private final Pagination pagination; // Missing pagination represents fetching all the records
-
+public abstract class Query {
```

Comment:
This was basically done to abstract out the internals of the Query class.
The clients should know nothing about the Query class. They should be able to simply build it and pass it on to `find` or `aggregate`.

Essentially, the new design enables us to change the QueryInternal structure the way we want.
Suppose, we have a new AggregationOperator (not exposed outside) that needs to be applied only by the Query Transformer, then, it's easy to make these changes in the QueryInternal class.
Basically, we are making the Query class an abstraction instead of an implementation. This way, the clients are agnostic of how the query class is used internally. For example, we may change the QueryInternal class to be mutable tomorrow (instead of building a whole new QueryInternal class) without affecting the client code. This also goes by the design principle that says "code for abstraction".

We don't have anything other than QueryInternal in mind as of now. But, this gives us a flexibility to do so if required in future.

Reply:
As discussed offline would be better to go with a concrete Query interface for method (find and aggregate)

Reply:
Done.